sys/xtimer/xtimer.c

```diff
@@ -80,63 +80,28 @@ void _xtimer_periodic_wakeup(uint32_t *last_wakeup, uint32_t period) {
 
     timer.callback = _callback_unlock_mutex;
     timer.arg = (void*) &mutex;
+    timer.target = timer.offset = timer.long_offset = 0;
 
-    uint32_t target = (*last_wakeup) + period;
+    /* time sensitive until setting offset */
+    uint8_t state = irq_disable();
     uint32_t now = _xtimer_now();
-    /* make sure we're not setting a value in the past */
-    if (now < (*last_wakeup)) {
-        /* base timer overflowed between last_wakeup and now */
-        if (!((now < target) && (target < (*last_wakeup)))) {
-            /* target time has already passed */
-            goto out;
-        }
-    }
-    else {
-        /* base timer did not overflow */
-        if ((((*last_wakeup) <= target) && (target <= now))) {
-            /* target time has already passed */
-            goto out;
-        }
-    }
+    uint32_t elapsed = now - (*last_wakeup);
+    uint32_t offset = (*last_wakeup) + period - now;
+    irq_restore(state);
```

Comment:
With the offset-based xtimer_core.c, this function's implementation becomes simpler like this. We only need to worry about the above time critical section where offset is calculated and the rest part becomes trivial.